### PR TITLE
feat(modeler): introduce the modeler api

### PR DESCRIPTION
**Pull request by daddykotex**

### CODE REVIEW

**Comment by daddykotex** [comment]
A few things annoy me, I'll let you review it and if we stumble on the same things, it probably means something is wrong.

**Comment by blemoine** [comment]
I would say that, strictly from the feature point of view, it seems OK.
But IMO, it's a little too Object Oriented (Interface + XImpl for everything).

File: spec/compression/compression.spec.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ describe('compression', function() {
     }, options));
 
     client.search.createBasicQuery(mql).then((response) => {
-      expect(response.rows.length).toBe(500);
+      expect(response.rows.length).toBeGreaterThanOrEqual(50);
```
Comment (jtheoof) [inline spec/compression/compression.spec.js]: Why the update of those tests?

File: spec/model/model.spec.js [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+* COPYRIGHT (c) 2017 Mnubo Inc. All Rights Reserved.
+*
+* The copyright to the computer program(s) herein is the property of Mnubo
+* Inc. The program(s) may be used and/or copied only with the written
+* permission from Mnubo Inc. or in accordance with the terms and conditions
+* stipulated in the agreement/contract under which the program(s) have been
+* supplied.
+*
+* ---------------------------------------------------------------------------
+*/
```
Comment (jtheoof) [inline spec/model/model.spec.js]: I don't know if the copyright applies for open source code. I'll check that.

File: src/model/model.ts [hunk 3]
```diff
@@ -64,15 +96,181 @@ export interface DataModel {
...
 
+export interface EntityOps<A> {
+  createOne(value: A): Promise<void>;
+  create(values: Array<A>): Promise<void>;
+  update(key: string, update: { displayName: string, description: string }): Promise<void>;
+
+  generateDeployCode(key: string): Promise<string>;
+  applyDeployCode(key: string, code: string): Promise<void>;
+  deploy(key: string): Promise<void>;
+}
```
Comment (jtheoof) [inline src/model/model.ts]: missing empty line after `}`

File: src/model/model.ts [hunk 4]
```diff
@@ -64,15 +96,181 @@ export interface DataModel {
   orphans: Orphans;
 }
 
+export interface EntityOps<A> {
+  createOne(value: A): Promise<void>;
+  create(values: Array<A>): Promise<void>;
```
Comment (jtheoof) [inline src/model/model.ts]: This is purely a preference of sdk user but I would have only one function 

`create(values: Array<A> | A)` and use `Array.isArray` to distinguish. What do you think?

File: spec/model/model.spec.js (hunk 2 above)
Comment (blemoine) [reply spec/model/model.spec.js]: Aside from the fact that the copyright is useless even in closed code (in Canada, as in most country in the world, the copyright is implied).

But here, with an Apache 2 license,  I'm sure that we have to remove it.

Comment (jtheoof) [reply spec/model/model.spec.js]: Agreed.

File: src/model/model.ts (hunk 4 above)
Comment (daddykotex) [reply src/model/model.ts]: It's a good idea

File: spec/model/model.spec.js (hunk 2 above)
Comment (daddykotex) [reply spec/model/model.spec.js]: I'll remove all of them

File: spec/compression/compression.spec.js (hunk 1 above)
Comment (daddykotex) [reply spec/compression/compression.spec.js]: The test is dependent on data in the namespace, and resets were issued on it, so there was not 500 events anymore

Comment (jtheoof) [reply spec/compression/compression.spec.js]: OK

**Comment by jtheoof** [comment]
LGTM, Anyone else?

**Comment by daddykotex** [comment]
I've updated the wiki to include some examples for the modeler API.

**Pull request merged**